contracts/Strategy.sol

```diff
@@ -348,70 +308,92 @@ contract Strategy is BaseStrategy {
+                lpts = uint(- deltas[deltas.length - 1]);
+                totalWantPooled += lpts;
             }
         }
         return totalWantPooled;
     }
 
-    function _getSwapRequest(IERC20 token, uint256 amount, uint256 lastChangeBlock) internal view returns (IBalancerPool.SwapRequest memory request){
-        return IBalancerPool.SwapRequest(IBalancerPool.SwapKind.GIVEN_IN,
-            token,
-            want,
-            amount,
-            balancerPoolId,
-            lastChangeBlock,
-            address(this),
-            address(this),
-            abi.encode(0)
-        );
+    /// Using a struct here to prevent stack too deep error
+    struct StackHelper {
+        uint8 assetsIndex;
+        uint8 swapsIndex;
+        bool isSame;
+        bool givenIn;
     }
 
-    function _sellBptForExactToken(uint256 _amountTokenOut) internal {
-        uint256[] memory amountsOut = new uint256[](numTokens);
-        amountsOut[tokenIndex] = _amountTokenOut;
-        bytes memory userData = abi.encode(IBalancerVault.ExitKind.BPT_IN_FOR_EXACT_TOKENS_OUT, amountsOut, balanceOfBpt());
-        IBalancerVault.ExitPoolRequest memory request = IBalancerVault.ExitPoolRequest(assets, minAmountsOut, userData, false);
-        balancerVault.exitPool(balancerPoolId, address(this), address(this), request);
+    /// Provide more flexibility during emergencies
+    function swap(uint _amount, IBalancerVault.SwapKind _swapKind, address _swap1Token, IBalancerPool _swap1Pool, address _swap2Token, IBalancerPool _swap2Pool, bool _mock) external onlyVaultManagers returns (int256[] memory _changes){
+        return _swap(_amount, _swapKind, _swap1Token, _swap1Pool, _swap2Token, _swap2Pool, false);
     }
 
-    // for partnership rewards like Lido or airdrops
+    /// In swaps of the type given_in you're pushing a token into a pipeline and getting another one at the end.
+    /// At each step along the way, the output of a swap becomes input of the next.
+    /// @param _amount = amount sent to pool to trade
+    /// In swaps of the type given_out you're pulling a token from a pipeline; the last step pulls some other token from your own account.
+    /// At each step along the way, the tokens that will go into a swap will be the ones that come out of the next.
+    /// @param _amount = amount desired to be returned by pool
+    function _swap(uint _amount, IBalancerVault.SwapKind _swapKind, address _swap1Token, IBalancerPool _swap1Pool, address _swap2Token, IBalancerPool _swap2Pool, bool _mock) internal returns (int256[] memory _changes){
+        if (_amount > 0) {
+            StackHelper memory sh;
+            sh.isSame = _swap1Token == address(poolData.lpt);
+
+            address[] memory assets = new address[](sh.isSame ? 2 : 3);
+            assets[sh.assetsIndex] = _swap1Token;
+            if (!sh.isSame) sh.assetsIndex++;
+            assets[sh.assetsIndex] = address(poolData.lpt);
+            sh.assetsIndex++;
```

Comment:
what happens if _swap1Token == _swap2Token

Reply:
weird things will probably happen lol. This is just meant to be used internally with pre-defined arguments, which wouldn't have `_swap1Token == _swap2Token`

Reply:
the external part is just if we ever need to rescue it from sms